feat: Add new Column Management Modal component

leSamo opened the pull request:
# Column Management Modal

Associated ticket: [RHINENG-3003](https://issues.redhat.com/browse/RHINENG-3003)

## What

Implement a modal that a user can use to configure the visibility of individual table columns. The modal supports columns that are hidden by default and also those that should be unhideable.

![image](https://github.com/patternfly/react-component-groups/assets/50696716/c5cadbea-fb3f-4a19-b55c-41f606f3fff1)

[Screencastfrom2024-04-1216-04-18-ezgif.com-video-to-gif-converter.webm](https://github.com/patternfly/react-component-groups/assets/8426204/268280f5-357e-48f1-a780-cb26b339bfef)

## Why
Since there is a request (see the associated ticket above) to implement column management in tables in all apps across Insights it would make sense to abstract a component that could be reused. Currently, the only implementation of column management is in [Insights Vulnerability](https://console.redhat.com/insights/vulnerability) (hidden inside kebab menu in table toolbar). This PR is based on the Vulnerability's implementation inspired by the [PF column management demo](https://www.patternfly.org/components/table/react-demos/column-management/).

## Example usage in practice with Redux

```js
// Hooks file
export const useColumnManagement = (columns, onApply) => {
    const [isModalOpen, setModalOpen] = useState(false);

    return [(
        <ColumnManagementModal
            appliedColumns={columns}
            applyColumns={newColumns => onApply(newColumns)}
            isModalOpen={isModalOpen}
            setModalOpen={setModalOpen}
        />), setModalOpen];
};

// Table file
const columns = useSelector(({ TableStore }) => TableStore.columns;
const [ColumnManagementModal, setColumnManagementModalOpen] = 
    useColumnManagement(columns, newColumns => dispatch(changeTableAction(newColumns)));

return (
    ...
    <Button onClick={() => setColumnManagementModalOpen(true)}>Manage columns</Button>
    ...
    { ColumnManagementModal }
    ...
)
```

patternfly-build commented:
Preview: https://react-component-groups-pr-component-groups-129.surge.sh

fhlavac commented on packages/module/src/ColumnManagementModal/ColumnManagementModal.tsx:
```diff
@@ -0,0 +1,146 @@
+    setModalOpen,
+    appliedColumns,
+    applyColumns,
+    ...props }: ColumnManagementModalProps) => {
```
Can you please add a default value of "ColumnManagementModal" to `ouiaId` property and also ouiaIds to important sub-elements (checkboxes, buttons)? We've just made some updates in OUIA ids across the components - see https://github.com/patternfly/react-component-groups?tab=readme-ov-file#testing and #131 
Thank you :-)

fhlavac commented on packages/module/src/ColumnManagementModal/ColumnManagementModal.tsx:
```diff
@@ -0,0 +1,146 @@
+      title="Manage columns"
+      onClose={() => setModalOpen(false)}
+      isOpen={isModalOpen}
+      variant="small"
```
```suggestion
      variant={ModalVariant.small}
```

fhlavac commented on packages/module/src/ColumnManagementModal/ColumnManagementModal.tsx:
```diff
@@ -0,0 +1,146 @@
+        <Button key="save" variant="primary" onClick={handleSave}>
+          Save
+        </Button>,
+        <Button key="cancel" variant="secondary" onClick={handleCancel}>
```
```suggestion
        <Button key="cancel" variant={ButtonVariant.link} onClick={handleCancel}>
```
This one should probably be a variant link as PatternFly recommends

fhlavac commented on packages/module/src/ColumnManagementModal/ColumnManagementModal.tsx:
```diff
@@ -0,0 +1,146 @@
+        </TextContent>
+      }
+      actions={[
+        <Button key="save" variant="primary" onClick={handleSave}>
```
```suggestion
        <Button key="save" variant={ButtonVariant.primary} onClick={handleSave}>
```

fhlavac commented on packages/module/src/ColumnManagementModal/ColumnManagementModal.tsx:
```diff
@@ -0,0 +1,146 @@
+          <Text component={TextVariants.p}>Selected categories will be displayed in the table.</Text>
+          <Split hasGutter>
+            <SplitItem>
+              <Button isInline onClick={selectAll} variant="link">
```
```suggestion
              <Button isInline onClick={selectAll} variant={ButtonVariant.link}>
```

fhlavac commented on packages/module/src/ColumnManagementModal/ColumnManagementModal.tsx:
```diff
@@ -0,0 +1,146 @@
+              </Button>
+            </SplitItem>
+            <SplitItem>
+              <Button isInline onClick={resetToDefault} variant="link">
```
```suggestion
              <Button isInline onClick={resetToDefault} variant={ButtonVariant.link}>
```

fhlavac commented on packages/module/src/ColumnManagementModal/ColumnManagementModal.tsx:
```diff
@@ -0,0 +1,146 @@
+
+  return (
+    <Modal
+      title="Manage columns"
```
With DataView, this component might be used also for other layouts than table, we could make this title a prop with a default value. What do you think?

fhlavac commented on packages/module/patternfly-docs/content/extensions/component-groups/examples/ColumnManagementModal/ColumnManagementModalExample.tsx:
```diff
@@ -0,0 +1,98 @@
+  }
+];
+
+/* eslint-disable camelcase */
```
```suggestion
```
Can we rename the props to camelcase to avoid unnecessary code in the examples. I'm also thinking if it's not better to simplify a bit the names of the cells to be more straightforward (ID, publish date, impact, score) - just to minimize the first-sight complexity of the example. But that's nitpicking

fhlavac commented on packages/module/patternfly-docs/content/extensions/component-groups/examples/ColumnManagementModal/ColumnManagementModalExample.tsx:
```suggestion
```
this one is probably not needed

fhlavac commented on packages/module/src/ColumnManagementModal/ColumnManagementModal.tsx:
```diff
@@ -0,0 +1,146 @@
+      ]}
+      {...props}
+    >
+      <DataList aria-label="Table column management" id="table-column-management" isCompact>
```
is it ok to replace the id with ouiaId?

fhlavac commented on packages/module/src/ColumnManagementModal/ColumnManagementModal.tsx:
```diff
@@ -0,0 +1,146 @@
+              />
+              <DataListItemCells
+                dataListCells={[
+                  <DataListCell key={'table-column-management-item' + index}>
```
replace the id with ouiaId?

fhlavac commented on packages/module/src/ColumnManagementModal/ColumnManagementModal.tsx:
```diff
@@ -0,0 +1,146 @@
+              <DataListItemCells
+                dataListCells={[
+                  <DataListCell key={'table-column-management-item' + index}>
+                    <label htmlFor={'checkbox-' + index}>
```
```suggestion
                    <label htmlFor={`checkbox-${index}`}>
```

fhlavac commented on packages/module/src/ColumnManagementModal/ColumnManagementModal.tsx:
```diff
@@ -0,0 +1,146 @@
+  isShown?: boolean,
+  /** Set to false if the column should be hidden initially */
+  isShownByDefault: boolean,
+  /** The checkbox will be checked and disabled, this is applicable to identifier columns which user shouldn't be able to hide */
+  isAlwaysShown?: boolean
```
Correct me if I am wrong, but from the code, this seems to only disable the checkbox, not ensure it is checked. Perhaps we could rename it to `isDisabled` or change the logic so that `isShown = true` doesn't have to be present

leSamo replied:
Yes, you are right; `isDisabled` in the context of column config could be understood as the disabling of the whole column, so I went with `isUntoggleable`, it's a bit of a mouthful, but I didn't want to go with `isToggleable` so that the default value would not be `true`. If you have better name for the param, I will gladly go with it.

fhlavac replied:
That's a fair point, thank you!

fhlavac commented on packages/module/patternfly-docs/content/extensions/component-groups/examples/ColumnManagementModal/ColumnManagementModal.md:
```diff
@@ -0,0 +1,30 @@
+
+### Showing and hiding of table columns
+
+Clicking the "Manage columns" button will open the column management modal. The "CVE ID" column is set to unhideable with `isAlwaysShown: true`, therefore its checkbox is disabled. "CVSS base score" column is set to be hidden by default. Always make sure to set `isShownByDefault` and `isShown` to the same boolean value in the initial state.
```
> Always make sure to set `isShownByDefault` and `isShown` to the same boolean value in the initial state.
Why is this necessary? 

I assume that the `isShownByDefault` value is some kind of shared preset for all users and given view (what is recommended and what the user sees when loads the view for the very first time), right? Then we can always prioritize the `isShown` value and use the default one only if `isShown` is not present or to reset the configuration.

leSamo replied:
Yes, your assumption is correct, we do prioritize `isShown` above `isShownByDefault`, the only consequence of breaking this rule:
> Always make sure to set isShownByDefault and isShown to the same boolean value in the initial state.

Would be that initial column selection after page load would differ from what is selected after clicking "Reset to default", since reset to default uses `isShownByDefault` and initial page load (and every subsequent opening of the modal) uses `isShown`.

fhlavac replied:
Ah, ok, that's fine then. Thank you 🙂

fhlavac reviewed (changes requested):
@leSamo thank you for your contribution. It looks really good! I've added some comments. Some of them may be worth discussion. Please let me know what do you think or if you need any help 🙂

andrew-ronaldson reviewed (changes requested):
The only thing I am noticing is the modal buttons. There was a decision on PF to make the Cancel button be a plain text button rather than the secondary button styling. I found that our Managed columns react demo was not updated so I created an issue to make that change on our PatternFly site.

_Screenshot is a v6 example but the Cancel button is shown accurately._
<img width="615" alt="Screenshot 2024-04-18 at 12 27 32 PM" src="https://github.com/patternfly/react-component-groups/assets/28658548/ff187cdb-b1d4-4297-bbcd-5e66cf44f4c4">

leSamo commented on packages/module/src/ColumnManagementModal/ColumnManagementModal.tsx:
```diff
@@ -0,0 +1,146 @@
+  /** Flag to show the modal */
+  isModalOpen: boolean,
+  /** Invoked when modal visibility is changed */
+  setModalOpen: (shouldOpen: boolean) => void
```
Okay, I replaced it, should I include it `ColumnManagementModalProps` interface as well? I am rewriting props inherited from `ModalProps`, but without this it won't show up in docs.

fhlavac replied:
Yeah, let's probably rewrite them.  @leSamo we should maybe also mention in the docs that the component accepts all regular ModalProps

leSamo commented on packages/module/src/ColumnManagementModal/ColumnManagementModal.test.tsx:
It's okay, I will address this in a follow-up PR

fhlavac replied:
Cool! Thank you!

leSamo commented:
Thank you @fhlavac for the comprehensive review, I have (hopefully) addressed all of your and @ andrew-ronaldson's comments.

fhlavac commented on packages/module/patternfly-docs/content/extensions/component-groups/examples/ColumnManagementModal/ColumnManagementModal.md:
```diff
@@ -0,0 +1,30 @@
+### Showing and hiding of table columns
+
+Clicking the "Manage columns" button will open the column management modal. The "ID" column should not be toggleable, therefore its checkbox is disabled with `isUntoggleable: true`. "Score" column is set to be hidden by default. Always make sure to set `isShownByDefault` and `isShown` to the same boolean value in the initial state.
+
```
```suggestion
For further customization, you can utilize all properties of the [modal component](/components/modal), with the exception of `ref`.
```

fhlavac commented on packages/module/patternfly-docs/content/extensions/component-groups/examples/ColumnManagementModal/ColumnManagementModal.md:
```diff
@@ -0,0 +1,30 @@
+
+### Showing and hiding of table columns
+
+Clicking the "Manage columns" button will open the column management modal. The "ID" column should not be toggleable, therefore its checkbox is disabled with `isUntoggleable: true`. "Score" column is set to be hidden by default. Always make sure to set `isShownByDefault` and `isShown` to the same boolean value in the initial state.
+For further customization, you can utilize all properties of the [modal component](/components/modal), with the exception of `ref`.
```
```suggestion
Clicking the "Manage columns" button will open the column management modal. The "ID" column should not be toggleable, therefore its checkbox is disabled with `isUntoggleable: true`. The "Score" column is set to be hidden by default. Always make sure to set `isShownByDefault` and `isShown` to the same boolean value in the initial state. For further customization, you can utilize all properties of the [modal component](/components/modal), except `ref` and `children`.
```

fhlavac commented:
@andrew-ronaldson can you please take a look? So this one can be merged

andrew-ronaldson commented:
Looks good! I think the only comment I have leans more into the design guidelines. The example given includes a primary button as the trigger for this modal. That's fine on its own but if the table has a toolbar with other actions we should only have one primary button on the page. In OCP we had these buttons to trigger the manage column modal. 

If I want to use this code in my product, can I easily change that primary button to a secondary or icon button? If that's the case I think we just need a bit of text pointing to design guidelines for buttons.

![Screenshot 2024-04-24 at 10 09 26 AM](https://github.com/patternfly/react-component-groups/assets/28658548/07ce2a7a-096c-4879-bfa5-517c1842ed4e)

leSamo commented:
@andrew-ronaldson I changed the example button variant to secondary and added an icon to the example. The button is not part of the component, since there are various ways to implement it, I think the majority of the time the button will be inside of a kebab menu. Is this okay?

![Screenshot from 2024-04-24 22-17-08](https://github.com/patternfly/react-component-groups/assets/8426204/3996d405-353d-4557-a0f2-e6ad1ba4a213)

andrew-ronaldson reviewed (approved):
LGTM, thanks!
